Add a GenericQuery to replace ~12 Query classes

tomalec opened the pull request:
_:warning: this PR targets a feature branch together multiple improvements to Analytics codebase_
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Implements part of https://github.com/woocommerce/automatewoo/issues/1238

Add a GenericQuery to reduce duplicated code in Query classes.
Also, it should be exposed to community extensions for reuse for Analytics customization.

The goal is to remove all(most) `Automattic\WooCommerce\Admin\API\Reports\**\Query` classes. But for backward compatibility with external extensions using those, we will keep them deprecated for a while.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

This PR does not introduce any new functionality, it changes the underlying code structure, to effectively maintain the same behavior.

1. Run automated tests
2. Smoke test Analytics functionality looking for regression issues

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [x] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->
Replace `Automattic\WooCommerce\Admin\API\Reports\*\Query` classes with a single `GenericQuery` class.

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

tomalec committed:
Update Reports unit tests to use GenericQuery

tomalec committed:
Fix query name in download stats controller

tomalec committed:
Add a GenericQuery

to reduce duplicated code in Query classes.
Also, to expose it to community extensions for reuse.

tomalec committed:
Abandon Customers\Stats\Query

use Customers\Stats\Query with a custom name instead

tomalec committed:
Use GenericQuery instead of duplicated Query classes

Deprecate abandoned classes.

tomalec committed:
Fix filter name interpolation for `woocommerce_analytics_{$snake_name}_select_query`

tomalec committed:
Fix linting errors

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

tomalec committed:
Declare strict_types for GenericQuery.php

tomalec committed:
Fix whitespace around strict_types directive

ianlin commented on plugins/woocommerce/src/Admin/API/Reports/GenericQuery.php:
```diff
@@ -0,0 +1,84 @@
+ * $report = new GenericQuery( $args, 'coupons' );
+ * $mydata = $report->get_data();
+ */
+class GenericQuery extends \WC_Object_Query {
```
This is a newly introduced class so we can add a `@since x.x.x` here.

tomalec replied:
Thanks for the suggestion added in https://github.com/woocommerce/woocommerce/pull/48964/commits/a0b16942a222e80a76acc10bd487ede96d10833b

ianlin commented on plugins/woocommerce/src/Admin/API/Reports/GenericQuery.php:
```diff
@@ -0,0 +1,84 @@
+		 *
+		 * @param array $query_args Query args.
+		 */
+		$args = apply_filters( "woocommerce_analytics_{$snake_name}_query_args", $this->get_query_vars() );
```
Those filters, for example `woocommerce_analytics_products_query_args`, I believe they were introduced before `9.0.0`. But it's hard to trace back all the actual versions for each report that include the filter, so I think leaving it as `9.0.0` would be fine.

tomalec replied:
Yep, I was trying to investigate that, but those filters were added so long ago and probably at different times for different reports, so it's hard to give one exact version.

ianlin commented on plugins/woocommerce/src/Admin/API/Reports/Categories/Query.php:
```diff
@@ -21,7 +21,9 @@
- * API\Reports\Query
+ * API\Reports\Categories\Query
+ *
+ * @deprecated 9.1 Categories\Query class is deprecated, please use GenericQuery or \WC_Object_Query instead.
```
I think `9.2` or `9.3` might be closer to the time we release this PR.

I tried replacing it with `x.x.x`, while the linter didn't complain, I'm not sure if it would be replaced to the release version when doing the release tho.

tomalec replied:
Agree, added in https://github.com/woocommerce/woocommerce/pull/48964/commits/a0b16942a222e80a76acc10bd487ede96d10833b

ianlin commented on plugins/woocommerce/src/Admin/API/Reports/Customers/Query.php:
```diff
@@ -16,14 +16,23 @@
+	 *
+	 * @var string
+	 */
+	public $name = 'customers';
```
In parent class `GenericQuery` the type of it is `protected`.

tomalec replied:
Fixed in https://github.com/woocommerce/woocommerce/pull/48964/commits/1d7339682062812c70ae6bd8d96a213463bf5a49

ianlin commented on plugins/woocommerce/src/Admin/API/Reports/Customers/Stats/Controller.php:
```diff
@@ -83,7 +85,7 @@ protected function prepare_reports_query( $request ) {
 	public function get_items( $request ) {
 		$query_args      = $this->prepare_reports_query( $request );
-		$customers_query = new Query( $query_args );
+		$customers_query = new Query( $query_args, 'customers-stats' );
```
I think we don't need to pass `customers-stats` to `Query`, as it's not using the new GenericQuery. Or maybe we should update `Customers/Stats/Query.php` to extends the new GenericQuery class?

tomalec replied:
We [do use here `CustomersQuery`](https://github.com/woocommerce/woocommerce/blob/add/generic-queue/plugins/woocommerce/src/Admin/API/Reports/Customers/Stats/Controller.php#L10:L10) (`Customers/Query.php`, no `/Stats`) that extends `GenericQuery`.
So we could use the same default query vars as Customers but set a different custom name for the data store and filters.

ianlin replied:
Ah I see. Didn't realise the `Customers/Stats/Controller` is using the query from `Customers/Query` instead of `Customers/Stats/Query`. Now I understand it's because the default query vars for `Customers/Query` and `Customers/Stats/Query` are identical so we can reuse the same Query class. Unlike `Orders/Query` and `Orders/Stats/Query`, their default query vars are different so both of them extends the GenericQuery and override the `get_default_query_vars()` method.

ianlin reviewed (commented):
Thanks for cleaning up the report classes. I tried to check the code for the first round and left some comments that we can discuss earlier. Will do the second round tomorrow.

tomalec committed:
Use placeholder for deprecated version,

as the number may change till we release the changes.

Address https://github.com/woocommerce/woocommerce/pull/48964/files#r1671850201

tomalec committed:
Make CustomersQuery::name protected

Address https://github.com/woocommerce/woocommerce/pull/48964#discussion_r1671926056

ianlin reviewed (approved):
Thanks for extra changes. The cleanup looks great and the Analytics page is working normally. LGTM.

I just did a bit of the search regarding `x.x.x` and it seems that it's not supported for WC core release:

- https://github.com/woocommerce/woocommerce/pull/39701#issuecomment-1832060739
- https://github.com/woocommerce/woocommerce/pull/38694#pullrequestreview-1488465548

So I think we would need to replace all the `x.x.x` to the current dev version `9.2.0`. Although it's weird to me that if we don't release it in `9.2.0`, we will need to change everything to `9.3.0` again. Or we can leave the `x.x.x` in this PR for now as it will be merged into a feature branch `dev/cleanup-analytics`, once we finish all the works there we can then decide which version we would like to replace.
